[SelectionDAG] Use unaligned store to legalize `EXTRACT_VECTOR_ELT` type when Stack is non-realignable

Nirhar opened the pull request:
This patch ports the commit a6614ec5b7c1dbfc4b847884c5de780cf75e8e9c to SelectionDAG TypeLegalization.

Fixes #98044

Nirhar commented:
@phoebewang @RKSimon @arsenm

phoebewang commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -3534,11 +3550,9 @@ SDValue DAGTypeLegalizer::SplitVecOp_EXTRACT_VECTOR_ELT(SDNode *N) {
   Align SmallestAlign = DAG.getReducedAlign(VecVT, /*UseABI=*/false);
```
Can we compare with `DAG.getSubtarget().getFrameLowering()->getStackAlign()` and use the smaller one here?

Nirhar replied:
Added!

phoebewang commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -3531,14 +3548,14 @@ SDValue DAGTypeLegalizer::SplitVecOp_EXTRACT_VECTOR_ELT(SDNode *N) {
-  auto PtrInfo = MachinePointerInfo::getFixedStack(MF, FrameIndex);
-  SDValue Store = DAG.getStore(DAG.getEntryNode(), dl, Vec, StackPtr, PtrInfo,
-                               SmallestAlign);
+  MachineMemOperand *StoreMMO = getStackAlignedMMO(
+      StackPtr, DAG.getMachineFunction(), VecVT.isScalableVector());
+  SDValue Store = DAG.getStore(DAG.getEntryNode(), dl, Vec, StackPtr, StoreMMO);
```
I assumed we only need to change `SmallestAlign`. Why do we still need these change?

Nirhar replied:
Sorry! I thought your case covers some extra conditions that I didn't exactly understand. I've reduced my patch to only your suggestion. A follow up question: Could this patch https://github.com/llvm/llvm-project/commit/a6614ec5b7c1dbfc4b847884c5de780cf75e8e9c also be rewritten in the form suggested by you ?

Nirhar replied:
Also your suggestion made me change many lit tests, can you verify if they are correct?

phoebewang replied:
The changes in lit test are functionally correct, but I'm not sure if we prefer to align them or not.

Revisit the original change, I think we also realigned the stack in the `foo2`, but it's not used. I guess some later optimizations may schedule the order of mov instructions. So that approach is a bit suboptimal to me.

@RKSimon which solution do you prefer?

RKSimon replied:
Unnecessarily losing alignment on vector spills to stack that we know is aligned isn't great - we'll likely to see perf reductions as soon as the stores cross cachelines (ymm/zmm stores in particular).

Shouldn't we be addressing any issues we have inside SelectionDAG::getReducedAlign?

Nirhar replied:
If we consider modifying `SelectionDAG::getReducedAlign` I see two locations where we can teach it to consider the Stack alignment:
1. Teach `TLI::getVectorTypeBreakdown` to clamp down to Stack Alignment. I see that we dont consider Stack Alignment inside the function. 
https://github.com/llvm/llvm-project/blob/462d084241616627be1ac2b967a7fcba9b0facfe/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L2479
2.  [My Preference] Modify these lines in `getReducedAlign`:
https://github.com/llvm/llvm-project/blob/462d084241616627be1ac2b967a7fcba9b0facfe/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L2482-L2487
and clamp down RedAlign2 to StackAlign if it exceeds it.

Which one is a better solution?

RKSimon replied:
Try (2) first?

Nirhar replied:
Have implemented (2) and have corrected lit tests to suit it as well. Please check @phoebewang @RKSimon .

phoebewang replied:
The change is similar to previous one, but I interpret @RKSimon 's question is we prefer to stack realignment for performance. Maybe I misunderstood it?

Nirhar replied:
I think these new changes would definitely force unaligned access to memory, where we do not need one. In my use case stack alignment was disabled. Perhaps I should go back to the first version of the patch that uses `getStackAlignedMMO` ?

phoebewang replied:
I think this is a mirror problem with #82130. The key is `no-realign-stack`. Can we clamp it down to StackAlign only under this attribute? We would solve all the problem then.

Nirhar replied:
Have updated the patch to only limit to the stack alignment, if the stack is not realignable. Please check now @phoebewang

arsenm reviewed (commented):
Reword title to explain why, this is trying to avoid introducing stack realignment

Commit:
[SelectionDAG] Use unaligned store to legalize `EXTRACT_VECTOR_ELT` type when Stack is non-realignable

This patch sets the alignment of store instructions generated during type
legalization of extractelement instruction, after considering stack
alignment, if the stack is not realignable.

Fixes #98044

phoebewang reviewed (approved):
LGTM, thanks!

Nirhar commented:
@phoebewang can you merge it on my behalf, I see that I dont have the option to merge

RSilicon commented:
"This leads to the generation of a vmovdqa instruction, which can lead to a General Protection Fault. Here is the link to the godbolt demo: https://godbolt.org/z/33h7YGc5K"

Given this, should this backported to 19.x or is that not a problem there?

phoebewang commented:
/cherry-pick

phoebewang commented:
> "This leads to the generation of a vmovdqa instruction, which can lead to a General Protection Fault. Here is the link to the godbolt demo: https://godbolt.org/z/33h7YGc5K"
> 
> Given this, should this backported to 19.x or is that not a problem there?

I think it's good to backport.

phoebewang commented:
/cherry-pick
